lib/core/proxy.c

```diff
@@ -398,6 +398,52 @@ static void copy_stats(struct rp_generator_t *self)
     self->src_req->proxy_stats.bytes_read.body = self->client->bytes_read.body;
 }
 
+static void copy_ssl_info(h2o_req_t *req, struct st_h2o_socket_ssl_t *ssl)
+{
+    h2o_socket_t s;
+    s.ssl = ssl;
+    req->proxy_stats.ssl.protocol_version = h2o_socket_get_ssl_protocol_version(&s);
+    req->proxy_stats.ssl.session_reused = h2o_socket_get_ssl_session_reused(&s);
+    req->proxy_stats.ssl.cipher = h2o_socket_get_ssl_cipher(&s);;
+    req->proxy_stats.ssl.cipher_bits = h2o_socket_get_ssl_cipher_bits(&s);
+    /* server name and session id are omitted since they are not static data */    
+}
+
+h2o_iovec_t h2o__proxy_log_ssl_protocol_version(h2o_req_t *req) 
+{
+    const char *s = req->proxy_stats.ssl.protocol_version;
+    return s != NULL ? h2o_iovec_init(s, strlen(s)) : h2o_iovec_init(NULL, 0);
+}
+
+h2o_iovec_t h2o__proxy_log_ssl_cipher(h2o_req_t *req) 
+{
+    const char *s = req->proxy_stats.ssl.cipher;
+    return s != NULL ? h2o_iovec_init(s, strlen(s)) : h2o_iovec_init(NULL, 0);
+}
+
+h2o_iovec_t h2o__proxy_log_ssl_session_reused(h2o_req_t *req) 
+{
+    switch (req->proxy_stats.ssl.session_reused) {
+    case 0:  return h2o_iovec_init(H2O_STRLIT("0"));
+    case 1:  return h2o_iovec_init(H2O_STRLIT("1"));
+    default: return h2o_iovec_init(NULL, 0);
+    }
+}
+
+h2o_iovec_t h2o__proxy_log_ssl_cipher_bits(h2o_req_t *req) 
+{
+    h2o_mem_pool_t *pool = &req->pool;
+    int bits = req->proxy_stats.ssl.cipher_bits;
+    if (bits != 0) {
+        char *s = (char *)(pool != NULL ? h2o_mem_alloc_pool(pool, char, sizeof(H2O_INT16_LONGEST_STR))
```

Comment:
We can use `&req->pool` directly here. `req` is a valid pointer and therefore `&req->pool` would never be NULL.